### ISSUE TITLE
Unexpected nullable warning passing null to an oblivious parameter

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#40751]
Encountered while annotating `CSharpCompilation`

``` csharp

            // Returns false if another compilation sharing this manager finished binding earlier and we should reuse its results.
            private bool CreateAndSetSourceAssemblyFullBind(CSharpCompilation compilation)
            {
...
                                InitializeNoLock(
                                    referencedAssembliesMap,
                                    referencedModulesMap,
                                    boundReferenceDirectiveMap,
                                    boundReferenceDirectives,
                                    explicitReferences,
                                    implicitReferenceResolutions,
                                    hasCircularReference,
                                    resolutionDiagnostics.ToReadOnly(),
                                    ReferenceEquals(corLibrary, assemblySymbol) ? null! : corLibrary, // unnessary suppression
                                    modules,
                                    moduleReferences,
                                    assemblySymbol.SourceModule.GetReferencedAssemblySymbols(),
                                    aliasesOfReferencedAssemblies,
                                    assemblySymbol.SourceModule.GetUnifiedAssemblies());
...
            }
```

**Comment by jcouv** [comment]
This is by-design with current type substitution rules. We're substituting a non-nullable reference type into a type parameter of an oblivious type, and that results in a non-nullable type.

**Comment by jcouv** [comment]
Pushed out to 16.7. I would like to think a bit more about this. The type substitution behavior is by-design, but it does make it hard to implement/override oblivious APIs which is unfortunate.
